### PR TITLE
Support GAN in TFPark

### PR DESCRIPTION
**Pull request by yangw1234**
Mnist unconditional gan result
generated_digital_numbers_w_gan_better

### CODE REVIEW

**Comment by yangw1234** [comment]
tensorflow_gan has two gan api:
1. GAN estimator, which pass all arguments all at once [gan_estimator](https://github.com/tensorflow/gan/blob/master/tensorflow_gan/examples/colab_notebooks/tfgan_tutorial.ipynb)
2. tensorflow_gan.train. [train](https://github.com/tensorflow/gan/blob/master/tensorflow_gan/examples/mnist/train_lib.py) In this case, users should follow a few steps
a. Specify the input to your networks. (e.g. create dataset)
b. Set up your generator and discriminator using a `GANModel`. (constructor arguments: g_fn and d_fn)
c. Specify your loss using a `GANLoss`. (constructor arguments: gan_model, g_loss_fn, d_loss_fn)
d. Create your train ops using a `GANTrainOps`. (constructor arguments: gan_loss, g_optimizer, d_optimizer)
e. Run your train ops. (tfgan.train(gan_train_ops))

**Comment by yangw1234** [comment]
I did not find other gan libraries.

**Comment by jason-dai** [comment]
> I did not find other gan libraries.

How about https://github.com/bstriner/keras-adversarial?

**Comment by yangw1234** [comment]
> > I did not find other gan libraries.
> 
> How about https://github.com/bstriner/keras-adversarial?

Cool! I'll check it out.

**Comment by jason-dai** [comment]
I think we need to support both Estimator-style (methods/functions) and Keras-style (models) API.

**Comment by yangw1234** [comment]
jenkins pass http://172.168.2.101:8080/view/ZOO-PR/job/ZOO-PR-Validation/2809/

File: pyzoo/zoo/tfpark/gan/gan_estimator.py [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+            return fn(*args, counter=counter)
+        else:
+            return fn(*args)
+
+    def train(self, dataset, end_trigger):
+
+        with tf.Graph().as_default() as g:
+
+            generator_inputs = dataset.feature_tensors
+            real_data = dataset.label_tensors
```
Comment (jason-dai) [inline pyzoo/zoo/tfpark/gan/gan_estimator.py]: instead of using `feature_tensors` and `label_tensors`, maybe it's clearer to just use `tensors[0]` and `tensors[1]`

File: pyzoo/zoo/tfpark/gan/gan_estimator.py [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+
+            grads = g_grads + d_grads
+
+            with tf.control_dependencies(grads):
+                increase_counter = tf.assign_add(counter, 1)
+
+            g_param_size = sum([np.product(g.shape) for g in g_grads])
+            with tf.Session() as sess:
+                sess.run(tf.global_variables_initializer())
+                optimizer = TFOptimizer(loss, GanOptimMethod(self._discriminator_optim_method,
```
Comment (jason-dai) [inline pyzoo/zoo/tfpark/gan/gan_estimator.py]: need to pass the steps to `GanOptimMethod`, no?

File: zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/net/GanOptimMethod.scala [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+
+import scala.reflect.ClassTag
+
+class GanOptimMethod[@specialized(Float, Double) T: ClassTag](
+          val dOptim: OptimMethod[T],
+          val gOptim: OptimMethod[T],
+          val dSteps: Int,
+          val gSteps: Int,
+          gParamSize: Int)(implicit ev: TensorNumeric[T])
+  extends OptimMethod[T] {
```
Comment (jason-dai) [inline zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/net/GanOptimMethod.scala]: I think you also need to implement `updateHyperParameter` which will invoke the `updateHyperParameter` of `dOptim` and/or `gOptim`?

File: pyzoo/zoo/tfpark/gan/gan_estimator.py (hunk 1 above)
Comment (yangw1234) [reply pyzoo/zoo/tfpark/gan/gan_estimator.py]: will be fixed in #1806

File: pyzoo/zoo/tfpark/gan/gan_estimator.py (hunk 2 above)
Comment (yangw1234) [reply pyzoo/zoo/tfpark/gan/gan_estimator.py]: will be fixed in #1806

File: zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/net/GanOptimMethod.scala (hunk 3 above)
Comment (yangw1234) [reply zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/net/GanOptimMethod.scala]: will be fixed in #1806
